### PR TITLE
Add campath

### LINKED ISSUE #62: HLAE camera path support
HLAE can import and export camera values from their own format .agr.
People can load the file in most 3D software rleady with HLAE own effort on supporting that.

Well, I just think if bxt-rs can load the file and record with sampling then bxt-rs can dilute a certain signature looking videomaking style.

All of the implementation should already been done in HLAE, my hope is that this would just be simple port.

### PR DESCRIPTION
**Pull request by khanghugo**
Closes #62 

Blender side must use HLAE Blender plugins: https://github.com/advancedfx/afx-blender-scripts.

The current state is very good despite some TODOs that I think are very niche. ~~I have not made anything with this to see if anything needs changing.~~

Lots of stuffs look verbose because I don't know Rust magic macros/built-in.

Both .Bvh and .Cam are supported is because .Bvh is more GoldSrc and .Cam is for Sourcn and not supported for GoldSrc by HLAE. I am just going extra to make sure anyone using it before could have it work here just fine. The difference is that .Bvh uses uniform frame time while .Cam uses timestamp and that works extremely well with Blender. It also has field of view for each entry so that enables some more effects.

The reason why we need export is to have a good reference of where the player is in Blender in order to aim better.

Default for offset is -0.25 because of loading time.

### CODE REVIEW

**Comment by khanghugo** [comment]
The current implementation will override view even if current demo time is earlier than what is in the demo. I wasn't really thinking about this while making the other video. It should work like that because that is how it does when campath file ends. Will be an easy addition.

File: src/modules/campath/mod.rs [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+            && engine::scr_fov_value.is_set(marker)
+            && engine::V_RenderView.is_set(marker)
+    }
+}
+
+static BXT_CAMPATH_LOAD: CVar = CVar::new(
+    b"bxt_campath_load\0",
+    b"\0",
+    "\
+Loads campath .cam or .bvh file and playback during demo.
```
Comment (YaLTeR) [inline src/modules/campath/mod.rs:63]:
```suggestion
Loads campath .cam or .bvh file for playback during demo.
```

File: src/modules/campath/mod.rs [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+",
+);
+
+static BXT_CAMPATH_OFFSET: CVar = CVar::new(
+    b"bxt_campath_offset\0",
+    b"-0.25\0",
+    "\
+Offsets time values in second of each campath entry. Offset can be negative. Default is -0.25.
```
Comment (YaLTeR) [inline src/modules/campath/mod.rs:71]: > Default for offset is -0.25 because of loading time.

Why? Is this somehow not recorded during recording but played back during playback?

File: src/modules/campath/mod.rs [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+        );
+    }
+}
+
+static BXT_CAMPATH_EXPORT_STOP: Command = Command::new(
+    b"bxt_campath_export_stop\0",
+    handler!(
+        "bxt_campath_export_stop
+        
+Stops capturing player motion. Automatically stops when demo stops.",
```
Comment (YaLTeR) [inline src/modules/campath/mod.rs:299]:
```suggestion
Stops capturing player motion.",
```
If anything, this should go into export_start docs.

File: src/modules/campath/common.rs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    (1. - t) * v0 + t * v1
+}
+
+pub fn angle_diff(a1: f64, a2: f64) -> f64 {
+    let a1 = a1.to_radians();
+    let a2 = a2.to_radians();
+
+    (a2.sin() * a1.cos() - a2.cos() * a1.sin())
+        .asin()
+        .to_degrees()
```
Comment (YaLTeR) [inline src/modules/campath/common.rs:34]: What does this do? Why does it compute asin(sin(a2 - a1)) in a roundabout way?

File: src/modules/campath/camio.rs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+        if time >= end.time {
+            return self.campaths.len();
+        }
+
+        // TODO: it can be more efficient
+        self.campaths
+            .iter()
+            .enumerate()
+            .find(|(_, e)| time < e.time)
```
Comment (YaLTeR) [inline src/modules/campath/camio.rs:40]: .position()

also you can combine it with condition above using .unwrap_or(self.campaths.len())

File: src/modules/campath/camio.rs [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+            fov,
+            viewinfo: ViewInfo {
+                vieworg: [xpos, ypos, zpos],
+                viewangles: [yrot, zrot, xrot], // be mindful of the flip
+            },
+        },
+    )(i)
+}
+
+pub fn read_camio(i: &str) -> IResult<&str, CamIO> {
```
Comment (YaLTeR) [inline src/modules/campath/camio.rs:131]: Please add some tests for the parsing. Check hltas for example.

File: src/modules/campath/common.rs [hunk 7]
```diff
@@ -0,0 +1,35 @@
+use nom::bytes::complete::tag;
+use nom::character::complete::space0;
+use nom::combinator::{map, opt, recognize};
+use nom::number::complete::float;
+use nom::sequence::preceded;
+use nom::IResult;
+
+#[derive(Clone, Copy)]
+pub struct ViewInfo {
+    pub vieworg: [f64; 3],
```
Comment (YaLTeR) [inline src/modules/campath/common.rs:10]: If that helps, we should have Vec3 and similar types, they probably have lerp on them and stuff.

File: src/modules/campath/exporter.rs [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+            .expect("Error: Cannot create .cam motion file.");
+        let mut file = BufWriter::new(file);
+
+        file.write_all(self.header().as_bytes())
+            .expect("Error: Cannot write motion into buffer.");
+        for idx in 0..self.data.campaths.len() {
+            file.write_all(self.entry_to_string(idx).as_bytes())
+                .expect("Error: Cannot write motion into buffer.");
+        }
+        file.flush()
```
Comment (YaLTeR) [inline src/modules/campath/exporter.rs:80]: Why is this needed?

File: src/modules/campath/bvh.rs [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+            cam_float, cam_float, cam_float, cam_float, cam_float, cam_float,
+        )), // parameters appear in the order of CHANNELS 6 specified in channel()
+        |(ypos, zpos, xpos, zrot, xrot, yrot)| ViewInfo {
+            vieworg: [-xpos, -ypos, zpos], // HLAE does this
+            viewangles: [-xrot, yrot, -zrot],
+        },
+    )(i)
+}
+
+pub fn read_bvh(i: &str) -> IResult<&str, Bvh> {
```
Comment (YaLTeR) [inline src/modules/campath/bvh.rs:125]: Also please add parsing tests.

**Review by YaLTeR — CHANGES_REQUESTED** [review]

File: src/modules/campath/mod.rs (hunk 2 above)
Comment (khanghugo) [reply src/modules/campath/mod.rs:71]: The loading time is when you load the campath as soon as the demo/map starts. When you record a campath, it is not necessary that you are recording as soon as the map starts. So this offset only happens when loading demo. It is more than often that campath is played during demo playback, so this offset is for that.

File: src/modules/campath/common.rs (hunk 4 above)
Comment (khanghugo) [reply src/modules/campath/common.rs:34]: Smaller angle between two vectors. I did not know I could run this through WA (I am not major in intellectual field)

File: src/modules/campath/exporter.rs (hunk 8 above)
Comment (khanghugo) [reply src/modules/campath/exporter.rs:80]: Is that not how it works? I really have no idea how this bufwriter works so I guess it writes into a buffer (in memory) then I need to flush it to write to file. My intention is to write into memory before writing into file because in case of export 60fps or more (10-15fps is my usual to-go).

Comment (YaLTeR) [reply src/modules/campath/exporter.rs:80]: I would expect it to flush on drop automatically. Can you check?

Comment (YaLTeR) [reply src/modules/campath/exporter.rs]: Sorry, I should have checked the docs myself, too.

> It is critical to call [flush](https://doc.rust-lang.org/std/io/struct.BufWriter.html#method.flush) before BufWriter<W> is dropped. Though dropping will attempt to [flush](https://doc.rust-lang.org/std/io/struct.BufWriter.html#method.flush) the contents of the buffer, any errors that happen in the process of dropping will be ignored. Calling flush ensures that the buffer is empty and thus dropping will not even attempt file operations.

In that case it is indeed better to have an explicit flush() with expect(). Could you please add it back? Otherwise the PR looks good.

**Comment by YaLTeR** [comment]
Thanks
